[WIP] Added support for resetting wind states to external observation

RomanBapst opened the pull request:
### Solved Problem
Initial work to support https://github.com/mavlink/mavlink/pull/2122

Once the mavlink message is available I need to fetch the data from the command and push it into the estimator.

### Changelog Entry
For release notes:
```
Feature: Added support for MAV_CMD_EXTERNAL_WIND_ESTIMATE
```

dagar commented on src/modules/ekf2/EKF/ekf.cpp:
```diff
@@ -346,6 +347,38 @@ void Ekf::resetGlobalPosToExternalObservation(double lat_deg, double lon_deg, fl
 
 	resetHorizontalPositionToExternal(pos_corrected, math::max(accuracy, FLT_EPSILON));
 }
+void Ekf::resetWindToExternalObservation(float wind_speed, float wind_direction, float wind_speed_accuracy, float wind_direction_accuracy)
```
Wind (including the full derivation update) can be entirely disabled at compile time.

```suggestion

#if defined(CONFIG_EKF2_WIND)
void Ekf::resetWindToExternalObservation(float wind_speed, float wind_direction, float wind_speed_accuracy, float wind_direction_accuracy)
```

dagar replied:
It might make sense to move all these misc wind pieces to a new file EKF/wind.cpp (like EKF/velocity_fusion.cpp, etc), then we could simply not include it in the build instead of including ifdefs.

dagar commented on src/modules/ekf2/EKF/wind.cpp:
```diff
@@ -0,0 +1,104 @@
+{
+#if defined(CONFIG_EKF2_AIRSPEED)
+	if (_control_status.flags.fuse_aspd && isRecent(_airspeed_sample_delayed.time_us, 1e6)) {
+		resetWindUsingAirspeed(_airspeed_sample_delayed);
```
It would be good to fully untangle this one and not even need to store `_airspeed_sample_delayed)`. A wind reset using airspeed can then happen from the airspeed control logic when that new sample is popped.

haumarco replied:
Then how would you handle it when the reset is requested from sideslip_fusion or drag_fusion. We could check in there if airspeed is enabled. If so, we set the corresponding flag, if not we reset to zero. But I dont know if there's a benefit in doing so compared to the current implementation.

bresch replied:
I would argue that sideslip or drag fusion shouldn't be allowed to call a reset to airspeed. They could either call a reset to 0 or just a covariance reset/bump

dagar commented:
There's a merge conflict with the mavlink submodule, but is this otherwise good to go?

<img width="491" alt="Screenshot 2024-07-01 at 9 48 52 PM" src="https://github.com/PX4/PX4-Autopilot/assets/84712/6df45e69-ea12-4322-aa57-d53ca6e46754">

RomanBapst commented:
@dagar I fixed the merge conflict. I did some testing and realized that even when wind deadreckoning, the wind soluation would always be pulled back to the previous solution after the reset because the horizontal velocity variances were too low.
I now added resetting those velocity variances along with the wind variances, as this allows the estimator to settle for the solution aligned with the newly set wind.

haumarco commented on src/modules/ekf2/EKF/wind.cpp:
```diff
@@ -0,0 +1,109 @@
+
+void Ekf::resetWindToExternalObservation(float wind_speed, float wind_direction, float wind_speed_accuracy, float wind_direction_accuracy)
+{
+	const float wind_speed_constrained = math::max(wind_speed, 0.0f);
```
An upper limit is not necessary? I don't know, just to want to make sure.

haumarco commented on src/modules/ekf2/EKF/python/ekf_derivation/derivation.py:
```diff
@@ -313,6 +313,23 @@ def compute_wind_init_and_cov_from_airspeed(
     P = P.subs({sideslip: 0.0})
     wind = wind.subs({sideslip: 0.0})
     return (wind, P)
+def compute_wind_init_and_cov_from_wind_speed_and_direction(
+        wind_speed: sf.Scalar,
+        wind_direction: sf.Scalar,
+        wind_speed_var: sf.Scalar,
+        wind_direction_var: sf.Scalar
+)-> (sf.V2, sf.M22):
+    wind = sf.V2(wind_speed * sf.cos(wind_direction), wind_speed * sf.sin(wind_direction))
+    J = wind.jacobian([wind_speed, wind_direction])
+
+    R = sf.M22()
+    R[0,0] = wind_speed_var
+    R[1,1] = wind_direction_var
+
+    P = J * R * J.T
+
+    return (wind, P)
+
```
Can you change the used variable of the jacobian to H here and in compute_wind_init_and_cov_from_airspeed()? Just to be consistent with the other derivations.

haumarco reviewed (commented):
Maybe a sanity check could be useful that the velocity estimate cant be completely destroyed after a wrong user input.

bresch commented on src/modules/ekf2/EKF/python/ekf_derivation/derivation.py:
```diff
@@ -313,6 +313,23 @@ def compute_wind_init_and_cov_from_airspeed(
+    wind = sf.V2(wind_speed * sf.cos(wind_direction), wind_speed * sf.sin(wind_direction))
+    J = wind.jacobian([wind_speed, wind_direction])
+
+    R = sf.M22()
+    R[0,0] = wind_speed_var
+    R[1,1] = wind_direction_var
```
```suggestion
    R = sf.Matrix.diag([wind_speed_var, wind_direction_var])
```

bresch replied:
@haumarco why not?

bresch commented on src/modules/ekf2/EKF/wind.cpp:
```diff
@@ -0,0 +1,109 @@
+{
+	const float wind_speed_constrained = math::max(wind_speed, 0.0f);
+
+	// wind direction is given as azimuth where wind blows FROM, we need direction where wind blows TO
```
Why is the interface like this? Why not directly the direction when the wind blows to?

haumarco replied:
Checked with @RomanBapst :
Most of the time in aerospace engineering and daily life, wind is defined from where it comes from. Therefore we handle the transformation to "wind to" in the backend of PX4.

bresch commented on src/modules/ekf2/EKF/wind.cpp:
```diff
@@ -0,0 +1,109 @@
+
+	resetWindTo(wind, wind_var);
+
+	// reset the horizontal velocity variances to allow the velocity states to be pulled towards
+	// a solution that is aligned with the newly set wind estimates
```
What if the velocity state is correct? We might not always want to do this I guess. And when do we really need to reset the velocity? Isn't that a sign that something else isn't working correctly?

RomanBapst replied:
@bresch The issue arose when you fly around in fixed wing with a wrong wind estimate, e.g. when taking off GPS denied. Then if you just trigger a wind reset, the wind states will be pulled back to the old solution, which is obviously not desired. I'm open to  other suggestions to fix the underlying issue.
Maybe @priseborough  can also tell us how you achieved this in your other EKF project.

RomanBapst commented on src/modules/ekf2/EKF/wind.cpp:
```diff
@@ -0,0 +1,109 @@
+
+	sym::ComputeWindInitAndCovFromWindSpeedAndDirection(wind_speed_constrained, wind_direction_rad, wind_speed_var, wind_direction_var, &wind, &P_wind);
+
+	const Vector2f wind_var = P_wind.diag();
```
That's a fair point. I think it was related to some issues I had and I did not change it back

dagar commented on src/modules/ekf2/EKF2.cpp:
```diff
@@ -534,6 +534,13 @@ void EKF2::Run()
 			}
+
+			if (vehicle_command.command == vehicle_command_s::VEHICLE_CMD_EXTERNAL_WIND_ESTIMATE) {
+				if (_ekf.control_status_flags().wind_dead_reckoning) {
```
I'd push the control status flags checks into the EKF backend.

haumarco commented:
We wanted to change the implementation to match the implementation of the position fusion-reset approach:
When having large uncertainties on the position, the introduced wind error gets pushed into the position which causes a jump in position. This seemed counter intuitive for the user.

We therefore have decided to only allow setting the wind manually while on ground. For a correction mid-air, the reset position functionality can be used, since it corrects the wind with the position fusion.

bresch commented on src/modules/ekf2/EKF/ekf.h:
```diff
@@ -521,6 +521,18 @@ class Ekf final : public EstimatorInterface
+	* @brief Resets the wind states to an external observation
+	*
+	* @param wind_speed The wind speed in m/s
+	* @param wind_direction The azimuth (from true north) from where the wind is flowing from in degrees (0 to 360)
```
This might be the mavlink convention, but it's now how we use it in EKF2. We should convert it to a "wind direction vector" (when receiving the message in ekf2.cpp) before sending it to the EKF backend

bresch commented on src/modules/ekf2/EKF/aid_sources/drag/drag_fusion.cpp:
```diff
@@ -57,6 +57,10 @@ void Ekf::controlDragFusion(const imuSample &imu_delayed)
 		if (_drag_buffer->pop_first_older_than(imu_delayed.time_us, &drag_sample)) {
 			fuseDrag(drag_sample);
+
+			if (!_aid_src_drag.fused && !_control_status.flags.fuse_aspd && !_control_status.flags.fuse_beta) {
```
I don't think we would like to reset as soon as a measurement spike is rejected. Also, `fuseDrag` returns directly if the mcoef and bcoef are not set (https://github.com/PX4/PX4-Autopilot/pull/23277/files#diff-5cf72a0fdf96e3bc1019002679b271a5ccf1b96106efa83240576205ec5893f8L79); in that case we also don't want to reset.
I think you would need to do a slightly more complicated start/stop logic if you want to call a reset in here

bresch commented on src/modules/ekf2/EKF/aid_sources/sideslip/sideslip_fusion.cpp:
```diff
@@ -64,16 +64,11 @@ void Ekf::controlBetaFusion(const imuSample &imu_delayed)
 			}
 
-			fuseSideslip(_aid_src_sideslip);
+			_control_status.flags.wind = true;
```
So even when `fuseSideslip` always returns `false` then `_control_status.flags.wind` is still `true`?

bresch commented on src/modules/ekf2/EKF/covariance.cpp:
```diff
@@ -200,15 +200,15 @@ void Ekf::predictCovariance(const imuSample &imu_delayed)
-		float wind_vel_nsd_scaled = math::constrain(_params.wind_vel_nsd, 0.f,
-					    1.f) * (1.f + _params.wind_vel_nsd_scaler * fabsf(_height_rate_lpf));
+	// wind vel: add process noise
+	if (!_external_wind_init) {
```
I see that `_external_wind_init` is automatically set to false when starting airspeed fusion (https://github.com/PX4/PX4-Autopilot/pull/23277/files#diff-7da169004af9436e30e9e918ddd7c7621b30de929ea72261ebaed2205fca8f3fR139) but what about drag and sideslip?

bresch commented on src/modules/ekf2/EKF/wind.cpp:
```diff
@@ -0,0 +1,109 @@
+	P.uncorrelateCovarianceSetVariance<State::wind_vel.dof>(State::wind_vel.idx, sq(_params.initial_wind_uncertainty));
+}
+
+void Ekf::resetWind()
```
let's replace https://github.com/Auterion/PX4_firmware_private/blob/4fa693ef267b61b8c97cc437a1ea4c0bfe23e2ef/src/modules/ekf2/EKF/aid_sources/sideslip/sideslip_fusion.cpp#L117 by `resetWindCov` and remove `resetWind` completely.
